### DISCUSSION TITLE
Write Unit-Tests for @RateLimiting usage

### DISCUSSION THREAD

**Discussion by mf-guse** [MarcGiffing/bucket4j-spring-boot-starter#351] [Q&A]
Hi everybody,

currently I'm having a very tough time trying to write unit tests if I used @RateLimiting annotation correctly, the fallback method and the skipCondition expressions are working correctly.

What I can't do is using @SpringBootTest in my unit test (too much stuff is started).

My general idea is to use JCache, i.e. by using Caffeine in the unit test. 

Testclass looks like

```
@ExtendWith(SpringExtension.class)
@EnableCaching
@ContextConfiguration(classes = {LoginApplicationRateLimitTestConfiguration.class})
@TestPropertySource("classpath:application-ratelimit.properties")
public class LoginApplicationRateLimitTest {}
```

Configuration class looks like

```
@EnableAspectJAutoProxy
@Import({ProxyCachingConfiguration.class, CacheAutoConfiguration.class, Bucket4jAopConfig.class, Bucket4jStartupCheckConfiguration.class, ServiceConfiguration.class, JCacheBucket4jConfiguration.class})
@TestConfiguration
public class LoginApplicationRateLimitTestConfiguration {
...

    @Bean
    public CacheManager cacheManager(Caffeine<Object, Object> caffeine) {
        CaffeineCacheManager cacheManager = new
                CaffeineCacheManager();
        cacheManager.setCaffeine(caffeine);
        return cacheManager;
    }

    @Bean
    Caffeine<Object, Object> caffeineSpec() {
        return Caffeine.newBuilder()
                .initialCapacity(10)
                .maximumSize(100)
                .expireAfterAccess(2, TimeUnit.MINUTES);
    }


}

```

and application-ratelimit.properties look like

```
spring.cache.cache-names=login_bucket
spring.cache.caffeine.spec=maximumSize=1000000,expireAfterAccess=3600s
spring.cache.jcache.provider=com.github.benmanes.caffeine.jcache.spi.CaffeineCachingProvider

bucket4j.enabled=true
bucket4j.filter-config-caching-enabled=false
bucket4j.cache-to-use=jcache
bucket4j.methods[0].name=login_rate_limit
bucket4j.methods[0].cache-name=login_bucket
bucket4j.methods[0].rate-limit.bandwidths[0].initial-capacity=1
bucket4j.methods[0].rate-limit.bandwidths[0].capacity=1
bucket4j.methods[0].rate-limit.bandwidths[0].time=10
bucket4j.methods[0].rate-limit.bandwidths[0].unit=minutes
bucket4j.methods[0].rate-limit.bandwidths[0].refill-speed=interval
```

But always something is missing: SyncCacheResolver, RateLimitAspect or something else.

What would be the most elegant way to get the Unit-Test up and running?

**Comment by MarcGiffing** [comment]
add debug=true to your configuration. Then take a look at the auto configuration classes in the system output. Is the Bucket4jAopConfig correctly configured? Is JCache correctly configured from the Spring Boot Starter? Is JCacheBucket4jConfiguration missing the CacheManager bean?

**Comment by mf-guse** [comment]
Thanks for the hints, I'll have a look into this. As soon I have a solution I'll add it to the caffeine example project.

**Comment by mf-guse** [comment]
I created a pull request regarding testing `@RateLimited` in unit tests.
